src/net.cpp

```diff
@@ -1823,6 +1823,13 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
     }
 
     const bool inbound_onion = std::find(m_onion_binds.begin(), m_onion_binds.end(), addr_bind) != m_onion_binds.end();
+
+    // Do not accept inbound connections from I2P peers already connected to us,
+    // as creating new tunnels is expensive.
+    if (addr.IsI2P() && AlreadyConnectedToAddress(addr, ConnectionType::INBOUND)) {
+        return;
+    }
```

Comment:
Isn't the tunnel already created at this point? Afaik the tunnel creation is handled by the i2p gateway (e.g. i2pd) not on our side.

I'm also not sure what we are protecting against here. Assuming tunnel creation is sufficiently DoSy (which it shouldn't be), this isn't enough to prevent attacks as an attacker could just not connect from the same address.